internal/langserver/handlers/command/module_list.go

```diff
@@ -0,0 +1,145 @@
+	DependentModules []terraformModule `json:"depmodules"`
+}
+
+func ModuleListHandler(ctx context.Context, args cmd.CommandArgs) (interface{}, error) {
+	modUri, ok := args.GetString("uri")
+	if !ok || modUri == "" {
+		return nil, fmt.Errorf("%w: expected module uri argument to be set", code.InvalidParams.Err())
+	}
+
+	if !uri.IsURIValid(modUri) {
+		return nil, fmt.Errorf("URI %q is not valid", modUri)
+	}
+
+	modPath, err := uri.PathFromURI(modUri)
+	if err != nil {
+		return nil, err
+	}
+
+	mm, err := lsctx.ModuleFinder(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	found, _ := mm.ModuleByPath(modPath)
+	if found.ModManifest == nil {
+		return nil, nil
+	}
+
+	list := parseModuleRecords(found.ModManifest.Records)
+
+	return moduleListResponse{
+		Modules:       list,
+		FormatVersion: moduleListVersion,
+	}, nil
+}
+
+func parseModuleRecords(records []datadir.ModuleRecord) []terraformModule {
+	modules := make(map[string]terraformModule)
+	for _, manifest := range records {
+		if manifest.Dir == "." {
+			// this is the current directory, which is technically a module
+			// skipping as it's not relevant in the activity bar (yet?)
+			continue
+		}
+
+		moduleName := manifest.Key
+		subModuleName := ""
+
+		// determine if this module is nested in another module
+		// in the currecnt workspace by finding a period in the moduleName
+		// is it better to look at SourceAddr and compare?
+		if strings.Contains(manifest.Key, ".") {
+			v := strings.Split(manifest.Key, ".")
+			moduleName = v[0]
+			subModuleName = v[1]
+		}
+
+		// github modules are put in `files` directory, but Manifest Record does not
+		// record the `name` in Key it puts it in SourceAddr
+		if strings.HasPrefix(manifest.Key, "files") {
+			moduleName = manifest.SourceAddr
+		}
+
+		// build what we know
+		moduleInfo := terraformModule{
+			Name:    moduleName,
+			Path:    strings.TrimSuffix(manifest.SourceAddr, "/"),
+			Version: manifest.VersionStr,
+			Type:    getModuleType(manifest),
+		}
+
+		if moduleInfo.Type == "tfregistry" {
+			moduleInfo.DocLink = getModuleDocumentationLink(manifest)
+		}
+
+		m, present := modules[moduleName]
+		if present {
+			// this module is located inside another so append
+			moduleInfo.Name = subModuleName
+			m.DependentModules = append(m.DependentModules, moduleInfo)
+			modules[moduleName] = m
+		} else {
+			// this is the first we've seen module
+			modules[moduleName] = moduleInfo
+		}
+	}
+
+	// don't need the map anymore, return a list of modules found
+	list := make([]terraformModule, 0)
+	for _, mo := range modules {
+		list = append(list, mo)
+	}
+
+	sort.SliceStable(list, func(i, j int) bool {
+		return list[i].Name < list[j].Name
+	})
+
+	return list
+}
+
+func getModuleType(manifest datadir.ModuleRecord) string {
+	if strings.HasPrefix(manifest.Dir, "modules") {
+		return "local"
+	} else if strings.HasPrefix(manifest.Key, "files") {
+		return "github"
+	} else {
+		return "tfregistry"
```

Comment:
I'm not sure we can make assumptions like above based on the `Dir` or `Key`. I think ideally the only input for detection should be just `manifest.SourceAddr`.

We may be able to reuse some of the logic here https://github.com/hashicorp/terraform/blob/332ea1f233891e45bd274e714ab8e4e11c79dc95/internal/addrs/module_source.go#L191-L268 and relevant tests https://github.com/hashicorp/terraform/blob/332ea1f233891e45bd274e714ab8e4e11c79dc95/internal/addrs/module_source_test.go

or alternatively/additionally the logic in go-getter detectors, which core uses [internally](https://github.com/hashicorp/terraform/blob/332ea1f233891e45bd274e714ab8e4e11c79dc95/internal/getmodules/getter.go#L31-L50):
 - https://github.com/hashicorp/go-getter/blob/main/detect_github.go
 - https://github.com/hashicorp/go-getter/blob/main/detect_git.go
 - https://github.com/hashicorp/go-getter/blob/main/detect_bitbucket.go
 - https://github.com/hashicorp/go-getter/blob/main/detect_gcs.go
 - https://github.com/hashicorp/go-getter/blob/main/detect_s3.go

Reply:
Used `go-getter` and `terraform-registry-address` to parse source addresses.